### PR TITLE
Change media endpoint structure

### LINKED ISSUE #245: Incorrect media path formatting
See:
https://github.com/dadi/api/blob/01190d602a9b1f82155fe868c7cf0a1c3e9e542b/dadi/lib/storage/disk.js#L37

Path is using `server.host`. It should not be using a local value and should instead use a new variable held in `media` block.

### LINKED ISSUE #260: Media uploads fail
All PUT requests (update) to the a collection of type "media" return a 404, despite containing a valid _id
They don't necessarily include an uploaded file, the issue happens with a standard text field update

API 1.16.3

### LINKED ISSUE #265: Arrays are not validated against schemas in POST request
To a schema without the field `foo`, a post body of `{"foo": ["5910a5352b36cf967a0fcd4b"]}` is stored against the collection and not identified as unavailable in schema.

### PR DESCRIPTION
**Pull request by eduardoboucas**
As per the conversation on Slack this morning, this PR introduces a few changes to how media is handled by API.

Currently, media assets are stored in media collections, which consist of normal collections with a version, database and collection schema, accessible on the normal REST endpoints (i.e. `/version/database/collectionName`). All that differentiates them from normal data collections is a `type` property in the `settings` block set to `media`.

With this PR, the concept of media collections is abstracted from the public API. It removes the requirement for a collection schema, using a schema kept internally in API. At the moment it's hardcoded to store images (containing dimensions, size, mime type, etc.), but in the future we can look into making the schema adapt to the type of file being uploaded.

## Endpoints

- Requesting a signed URL for a media upload:

  ```
  POST /media/sign
  ```

- Uploading a media asset

  ```
  POST /media/sign/:signedUrl
  ```

- Listing media assets

  ```
  GET /media
  ```

- Accessing a specific media asset

  ```
  GET /media/:assetPath (e.g. `/media/2017/04/27/flowers.jpg`)
  ```

 ## Media buckets

Even though that's abstracted from the end user, assets still need to be stored in collections. Assets POSTed to `/media` will be stored in a `mediaStore` collection (configurable via the `media.defaultBucket` configuration parameter). We can have as many namespaces (what we called "media buckets") as we want. That's configurable as an array in the `media.buckets` parameter (defaulting to `[]`).

This is how the endpoints look like for interacting with a media bucket called `mediaAvatars`:

- Requesting a signed URL for a media upload:

  ```
  POST /media/mediaAvatars/sign
  ```

- Uploading a media asset

  ```
  POST /media/mediaAvatars/sign/:signedUrl
  ```

- Listing media assets

  ```
  GET /media/mediaAvatars
  ```

- Accessing a specific media asset

  ```
  GET /media/mediaAvatars/:assetPath (e.g. `/media/mediaAvatars/2017/04/27/flowers.jpg`)
  ```

## Naming conflicts

Whenever there is a data collection with the same name as one of the media buckets, API throws an error detailing the name of the conflicting collection.

## API base URL

This also fixes the issue described in #245 by introducing a `server.baseUrl` configuration parameter, which will be used to determine the URL of the media assets when using the disk storage option.

---

@mingard can you confirm whether this meets the requirements?

@jimlambie could you share your thoughts on the approach/implementation? Happy to make any required changes.

Thanks!

Fix #260 
Fix #245 
Fix #265

### CODE REVIEW

**Comment by mingard** [comment]
@eduardoboucas Pagination, searching by `_id` and filtering work?

**Comment by eduardoboucas** [comment]
Pagination and filtering are working. Searching by `_id` hasn't been implemented from the beginning, I think because of potential conflicts with the asset delivery endpoint. I'll check with @jimlambie though.

**Comment by eduardoboucas** [comment]
At the moment you can still search by ID using a filter (e.g. `/media?filter={"_id":"5901e528724fc24c8784dfb9f"}`, just not the ID endpoint.

**Comment by mingard** [comment]
@eduardoboucas great, then this is ready.

**Comment by eduardoboucas** [comment]
As mentioned in #245, we can solve 1 and 2 by moving `server.baseUrl` to its own block, following the same convention as used in `server`:

```json
"baseUrl": {
  "protocol": "http",
  "port": 80,
  "host": "mydomain.com"
}
```

3. I'll change that. We'll store the relative path and make API prepend the base URL before returning the documents.

4. I can remove this from the output before returning the documents

5. I'll look into encoding the client ID as part of the signed URL token.

**Comment by mingard** [comment]
@eduardoboucas above sounds fine. In addition, should `protocol` be stored? Could be substitute this for an **ssl** Boolean?

**Comment by eduardoboucas** [comment]
> @eduardoboucas above sounds fine. In addition, should protocol be stored? Could be substitute this for an ssl Boolean?

We could, but I think it's important that we keep consistency between the various config blocks, and `server` uses a `protocol` property. If we decide to change it, we should change it in both places, but that should be left for v2.0 as it's a breaking change.

**Comment by jimlambie** [comment]
The configuration property `url` isn't descriptive enough for me. Could we use something like `publicUrl` or similar? Thoughts?

**Comment by eduardoboucas** [comment]
> The configuration property url isn't descriptive enough for me. Could we use something like publicUrl or similar? Thoughts?

Agree 100%. `publicUrl` works for me.

File: dadi/lib/composer/index.js [hunk 1]
```diff
@@ -89,6 +83,13 @@ Composer.prototype.composeOne = function (doc, callback) {
             }
           }
 
+          // Are we composing a media document? If so, we need to format it
```
Comment (eduardoboucas) [inline dadi/lib/composer/index.js]: @jimlambie I don't like this. It feels like we should have the ability to have built-in hooks (as in not actual files on `workspace/hooks`) that would process documents. In this case, a hook would check whether a document is a media asset and, if so, remove some of the properties and change the relative path to a full URL.

**Comment by eduardoboucas** [comment]
Added acceptance tests covering listing media assets, requesting a signed URL and uploading an asset.

Also solved an issue where relative URLs were not being resolved on `POST`.

File: utils/create-client.js [hunk 2]
```diff
@@ -42,7 +42,7 @@ connection.on('connect', db => {
       },
       {
         label: '-> Access type (admin, user)',
-        key: 'type',
+        key: 'accessType',
```
Comment (eduardoboucas) [inline utils/create-client.js]: @jimlambie This should be `accessType` and not `type`, amirite?

**Comment by eduardoboucas** [comment]
Changed `url` config block to `publicUrl`, as per discussion.

File: dadi/lib/composer/index.js (hunk 1 above)
Comment (jimlambie) [inline dadi/lib/composer/index.js]: @eduardoboucas I don't particularly like it either. Internal hooks are a sound idea. Agree that's it's worth revisiting at some point!

File: utils/create-client.js (hunk 2 above)
Comment (jimlambie) [inline utils/create-client.js]: urrite!

**Comment by jimlambie** [comment]
Fix #260

**Comment by eduardoboucas** [comment]
At the moment, media documents are being stored with the relative path on the `path` property (e.g. `"/media/2017/05/16/3RdYMTLoL1X16djGF52cFtJovDT-1494966057926.jpg"`), which is transformed into a full URL when the document is returned.

I propose that instead of replacing the contents of `path`, we leave that as it is and write the full URL to a new property called `url`. We can still use `url` to render the image directly, but `path` is useful if we want to serve the images through CDN, without having to manipulate the URL.

This is how it'd look like:

```json
"image": {
    "_id": "591b5f29795b683664af01e9",
    "fileName": "3RdYMTLoL1X16djGF52cFtJovDT.jpg",
    "mimetype": "image/jpeg",
    "width": 600,
    "height": 900,
    "contentLength": 54907,
    "path": "/media/2017/05/16/3RdYMTLoL1X16djGF52cFtJovDT-1494966057926.jpg",
    "createdAt": 1494966057685,
    "createdBy": null,
    "v": 1,
    "url": "http://localhost:5000/media/2017/05/16/3RdYMTLoL1X16djGF52cFtJovDT-1494966057926.jpg"
}
```

(@mingard this is a one-line change in Publish, which I've done on my local instance and that will commit if we're happy with this.)

**Comment by eduardoboucas** [comment]
@jimlambie We have some merge conflicts here. Could this be because you progressed the previous media implementation further? Could you help me sort this out so I don't break the internet? 😇

**Comment by eduardoboucas** [comment]
I propose that we add information about the media buckets to the `/api/collections` endpoint, indicating a list of the available media buckets as well as the name of the default one. Something like this:

```
GET /api/endpoints

{
  "collections": [
    {
      "version": "1.0",
      "database": "publish",
      "name": "Articles",
      "slug": "articles",
      "path": "/1.0/publish/articles"
    },
    {
      "version": "1.0",
      "database": "publish",
      "name": "Books",
      "slug": "books",
      "path": "/1.0/publish/books"
    },
    {
      "version": "1.0",
      "database": "publish",
      "name": "Publish authors",
      "slug": "users",
      "path": "/1.0/publish/users"
    }
  ],
  "media": {
    "buckets": [
      "userAvatars",
      "mediaStore"
    ],
    "defaultBucket": "mediaStore"
  }
}
```

**Comment by jimlambie** [comment]
Close #156

**Comment by jimlambie** [comment]
Close #156

**Pull request merged**